### PR TITLE
MOD-11416: Fix `total_result` Accuracy When Using FILTER

### PR DESCRIPTION
**Pull request by kei-nan**
This is both for FT.AGGREGATE and FT.HYBRID
When using FILTER the total_result count was not decremented.
This is essentially a bug fix.

A clear and concise description of what the PR is solving, including:
1. Current: Client receives an incorrect total_result when using FILTER
2. Change: Decrement the total results if the result was filtered
3. Outcome: Better accuracy when returning the number of total results to the user.

#### Main objects this PR modified
1. Filter result processor

#### Mark if applicable

- [X] This PR introduces API changes
- [ ] This PR introduces serialization changes

---

> [!NOTE]
> Ensure `total_results` reflects post-filtered outputs; update helper and tests to read/verify the correct count.
> 
> - **Backend**:
>   - **Filter RP**: In `src/aggregate/expr/expression.c` (`rpevalNext_filter`), decrement `rp->parent->totalResults` when a result is filtered (with assertion) so `total_results` matches emitted results.
> - **Tests/Utils**:
>   - **Helper**: `tests/pytests/common.py#get_results_from_hybrid_response` now returns `(results, total_results)` by parsing `total_results`.
>   - **Hybrid tests**: Update usages to unpack `(results, count)` and assert accurate `total_results`; fix expectations where filters reduce result count (e.g., post-filter from `3` to `1`).
>   - **GroupBy**: Add validation for filtered groups and `total_results`; introduce no-results case and consistency checks.
>   - Minor: remove debug prints; clarify skip comments for cluster.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b030321ecc5aff64c0307e48eda58aee5edd1796. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/aggregate/expr/expression.c [hunk 1]
```diff
@@ -479,6 +479,8 @@ static int rpevalNext_filter(ResultProcessor *rp, SearchResult *r) {
       return RS_RESULT_OK;
     }
 
+    // Reduce the total number of results
+    rp->parent->totalResults--;
```
Comment (oshadmi) [inline src/aggregate/expr/expression.c:484]: Cannot wrap, right?

File: tests/pytests/test_hybrid_apply_filter.py [hunk 2]
```diff
@@ -61,8 +61,9 @@ def test_hybrid_apply_filter_linear():
     query_vector = np.array([0, 0]).astype(np.float32).tobytes()
     response = env.cmd('FT.HYBRID', 'idx', 'SEARCH', 'green', 'VSIM' ,'@embedding', query_vector,\
          'COMBINE', 'LINEAR', '4', 'ALPHA', '0.0', 'BETA', '1.0', 'APPLY', '2*@__score', 'AS', 'doubled_score', 'FILTER', '@doubled_score>1')
-    results = get_results_from_hybrid_response(response)
+    results, count = get_results_from_hybrid_response(response)
     env.assertTrue(set(results.keys()) == {"doc:1"})
+    env.assertEqual(count, 1)
```
Comment (oshadmi) [inline tests/pytests/test_hybrid_apply_filter.py:66]: Can we add a test that reaches zero due to filter?

File: tests/pytests/test_hybrid_apply_filter.py [hunk 3]
```diff
@@ -80,5 +81,6 @@ def test_hybrid_apply_filter_rrf():
     response = env.cmd('FT.HYBRID', 'idx', 'SEARCH', search_query, 'VSIM' ,'@embedding', query_vector,\
         'COMBINE', 'RRF', '4', 'CONSTANT', '60', 'WINDOW', '10',
          'APPLY', '2*@__score', 'AS', 'doubled_score', 'FILTER', f'@doubled_score>{threshold - epsilon}')
-    results = get_results_from_hybrid_response(response)
+    results, count = get_results_from_hybrid_response(response)
     env.assertTrue(set(results.keys()) == {"doc:4"})
+    env.assertEqual(count, 1)
```
Comment (oshadmi) [inline tests/pytests/test_hybrid_apply_filter.py:86]: Can we add a test that reaches zero due to filter?

**Review by Itzikvaknin — APPROVED** [review]
Nice ⚡ 
I think we need a `response_get()` to replace 
```
    res_results_index = recursive_index(response, 'total_results')
    res_results_index[-1] += 1
    access_nested_list(response, res_results_index)
```
 But thats for another PR

File: src/aggregate/expr/expression.c (hunk 1 above)
Comment (kei-nan) [reply src/aggregate/expr/expression.c:484]: added an assert, don't think it is possible for it to wrap since it is tightly tied to the amount of results the loop is processing.

[the remark this answers is not in the document]
Comment (kei-nan) [reply src/aggregate/expr/expression.c:484]: I don't think it can underflow but will add an if and assert to ensure that

File: src/aggregate/expr/expression.c [hunk 4]
```diff
@@ -479,6 +479,9 @@ static int rpevalNext_filter(ResultProcessor *rp, SearchResult *r) {
       return RS_RESULT_OK;
     }
 
+    // Reduce the total number of results
+    RS_ASSERT(rp->parent->totalResults > 0);
+    rp->parent->totalResults--;
```
[the remark this answers is not in the document]
Comment (kei-nan) [reply src/aggregate/expr/expression.c:484]: I can't seem to please cursor, previously it commented that the additional if was redundant if we had an assertion.
I think the assertion is good enough.
resolving

File: tests/pytests/test_hybrid.py [hunk 5]
```diff
@@ -580,15 +580,15 @@ def test_knn_post_filter(self):
...
         expected = [
-            'total_results', 3,
             'results',
             [
                 ['__key', 'both_01', '__score', '0.45']
             ],
+            'total_results', 1,
             'warnings', [],
             'execution_time', ANY
         ]
```
Comment (GuyAv46) [inline tests/pytests/test_hybrid.py]: fix comments and case as well

File: tests/pytests/test_hybrid_apply_filter.py [hunk 6]
```diff
@@ -80,5 +81,26 @@ def test_hybrid_apply_filter_rrf():
...
+
+# TODO: remove once FT.HYBRID for cluster is implemented
```
Comment (GuyAv46) [inline tests/pytests/test_hybrid_apply_filter.py]:
```suggestion
# TODO: remove skip once FT.HYBRID for cluster is implemented
```

**Review by GuyAv46 — COMMENTED** [review]
About the description:
The counter is meaningless for `FT.AGGREGATE`, and this is probably why the `FILTER` didn't care about this counter. I'm not sure what it means for `FT.HYBRID`

**Review by oshadmi — APPROVED** [review]

File: tests/pytests/test_hybrid.py [hunk 7]
```diff
@@ -580,15 +580,15 @@ def test_knn_post_filter(self):
 
         # total_results should be the number of results before filtering.
```
Comment (nafraf) [inline tests/pytests/test_hybrid.py]: This comment should be updated, now the number of results is equal to the number of results after filtering

File: tests/pytests/test_hybrid.py [hunk 8]
```diff
@@ -580,15 +580,15 @@ def test_knn_post_filter(self):
...
+        self.env.assertEqual(filtered_dict['total_results'], 1)
 
         # But only 1 result is returned by the filtered query:
         expected = [
-            'total_results', 3,
             'results',
             [
                 ['__key', 'both_01', '__score', '0.45']
             ],
+            'total_results', 1,
```
Comment (oshadmi) [inline tests/pytests/test_hybrid.py]: Why change the order?
Also not aligned with #6889

File: src/aggregate/expr/expression.c (hunk 4 above)
Comment (GuyAv46) [inline src/aggregate/expr/expression.c:484]: We should have a test where the filter is after a grouping step, to make sure the count is still correct in this case

File: tests/pytests/test_hybrid_apply_filter.py (hunk 6 above)
Comment (kei-nan) [reply tests/pytests/test_hybrid_apply_filter.py]: replaced in all the comments

File: src/aggregate/expr/expression.c (hunk 4 above)
Comment (kei-nan) [reply src/aggregate/expr/expression.c:484]: added `test_hybrid_groupby_with_filter` in `test_hybrid_groupby.py`

**Comment by kei-nan** [comment]
> About the description:
> The counter is meaningless for `FT.AGGREGATE`, and this is probably why the `FILTER` didn't care about this counter. I'm not sure what it means for `FT.HYBRID`

So in hybrid the hybrid merger should update the total count of results when there is sorting involved.
We just need to ensure the count correctness in the future result processors - like filter.

**Review by oshadmi — APPROVED** [review]
